### DISCUSSION TITLE
[RFC] How do we do pan and zoom?

### DISCUSSION THREAD

**Discussion by srallen** [zooniverse/front-end-monorepo#2427] [Q&A]
We have four implementations of pan and zoom functionality in the monorepo. In addition, some CFEs have implemented this and may have unique requirements that we should be aware of as we get to parity function for projects to migrate over. Let's discuss the various ways of how we pan and zoom and ideally come to a decision that we can document as an ADR.

The four ways in the monorepo are:

1. PH TESS's `LightCurveViewer`: This is purely done in d3, coupled with the functionality required for PH TESS.
2. `SVGPanZoom`: This is essentially a port of PFE's functionality using a view box and currently used by the image viewers. This has to be wrapped around an svg element to be functional.
3. `VXZoom`: An extension of the vx library's pan and zoom function. Uses a transformation matrix and has some configurability. Currently implemented with the `ScatterPlotViewer` and subsequently the `VariableStarViewer`. Unknown if it is usable with image viewers, but in theory it is.
4. `SubjectGroupViewer`: A variant of `SVGPanZoom` since the transformations have to be applied to multiple image subject simultaneously.

**Comment by eatyourgreens** [comment]
We've also got the [`withKeyZoom`](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/withKeyZoom/withKeyZoom.js) decorator which takes `onPan`, `ZoomIn` and `ZoomOut` functions as props, and wires those up to `keydown` events emitted by the decorated component.

In theory, it should work with any implementation that can pass those three callback functions.

**Comment by eatyourgreens** [comment]
I think my vote would go to a component that can be wrapped around any SVG element (including groups of embedded SVGs for the group viewer) and passes down a transformation that operates on its children. Plus some mechanism for getting the image pixel coordinates of clicks when you are drawing on a photo. At the moment, we invert the current transformation matrix in order to do that. I think ASM might have a better implementation of that than the PFE implementation, but it's been a long time since I looked at that code.

**Comment by srallen** [comment]
@eatyourgreens vx has a utility function for getting point coordinates in events. It's unfortunately undocumented, but you can [read the code here](https://github.com/hshoff/vx/tree/master/packages/vx-event/src). I used it with `VXZoom` for a zoom to point function: 

https://github.com/zooniverse/front-end-monorepo/blob/8c310a7884dae858d5eaaae030dff82a2965ed11/packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SVGComponents/VXZoom/VXZoom.js#L50-L55

This might work?
